### PR TITLE
Fiber segmentation

### PR DESCRIPTION
**Pull request by ackagel**
## **Purpose**

Adds fiber segmentation capabilities to the repo.

## **Implementation**

Currently, only a simple function has been added, but the contents of a pending design doc will be added soon.

## **Remaining issues**

- [x] Notebook, tests, visualizations.
- [x] notebook successfully ran with the example dataset

### CODE REVIEW

**Comment by ackagel** [comment]
@angelolab/computational_team just want some basic feedback on the general structure of the main fiber segmentation function.  Further changes will probably come from a different pr/branch merging into this one.

@ngreenwald I can draft a design doc for the notebook/visualizations that we'll want for this stuff later, but I just wanted to get some of the cleaned up fiber code up sooner than later

**Review by alex-l-kong — COMMENTED** [review]
Looks good to me at first glance. I think some comments explaining some steps and providing more detailed motivations behind them would be good.

**Review by ngreenwald — CHANGES_REQUESTED** [review]
Agreed, first pass looks good. Only major code structure change will be making this batch compatible. 

For the notebook, I think the first section should be guidance on how the post-processing can be tweaked. The user selects a single image, and then each step is plotted for that single image, showing how increasing and decreasing the relevant parameter affects the results. Once they're happy with the outputs, the second half of the notebook is batch processing the entire cohort.

**Review by srivarra — COMMENTED** [review]
It looks good. The only thing I can think of is having an option to not save the `fiber_object_table` as a `csv` if for whatever reason it's not needed.

**Comment by ackagel** [comment]
@ngreenwald how many FoVs should I add for the example dataset?  I was thinking 5 and having the default batch number be 3, as a "make sure the batching is working" measure.

**Comment by ngreenwald** [comment]
Sure, that sounds good

File: ark/segmentation/fiber_segmentation_test.py [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+
+import ark.settings as settings
+from ark.utils import io_utils
+from ark.segmentation import fiber_segmentation
+
+
+def test_plot_fiber_segmentation_steps():
+    img_dir = os.path.join(Path(__file__).parent.parent.parent,
+                           "data", "example_dataset", "fiber_segmentation")
+
```
Comment (ngreenwald) [inline ark/segmentation/fiber_segmentation_test.py:14]: For tests, we can either use fake data, or the example dataset Sri added

File: ark/segmentation/fiber_segmentation_test.py [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+    # test success
+    fiber_segmentation.plot_fiber_segmentation_steps(img_dir, 'fov1', 'Collagen1')
+
+
+def test_run_fiber_segmentation():
+    with tempfile.TemporaryDirectory() as temp_dir:
+
+        img_dir = os.path.join(Path(__file__).parent.parent.parent,
+                               "data", "example_dataset", "fiber_segmentation")
+
```
Comment (ngreenwald) [inline ark/segmentation/fiber_segmentation_test.py:37]: Same thing here

File: ark/segmentation/fiber_segmentation.py [hunk 3]
```diff
@@ -0,0 +1,298 @@
...
+            imsave(os.path.join(debug_path, f'{fov}_thresholded.tiff'), threshed,
+                   check_contrast=False)
+            imsave(os.path.join(debug_path, f'{fov}_ridges_thresholded.tiff'),
+                   distance_transformed, check_contrast=False)
+            imsave(os.path.join(debug_path, f'{fov}_meijering_filter.tiff'), ridges,
+                   check_contrast=False)
+            imsave(os.path.join(debug_path, f'{fov}_contrast_adjusted.tiff'), contrast_adjusted,
+                   check_contrast=False)
+
+        imsave(os.path.join(out_dir, f'{fov}_fiber_labels.tiff'), labeled_filtered,
```
Comment (ngreenwald) [inline ark/segmentation/fiber_segmentation.py:283]: Not sure if this is supposed to be indented or not. If not, why do we also return the dictionary?

File: ark/utils/plot_utils.py [hunk 4]
```diff
@@ -371,6 +371,28 @@ def create_overlay(fov, segmentation_dir, data_dir,
     return rescaled
 
 
+def set_minimum_color_for_colormap(cmap, default=(0, 0, 0, 1)):
```
Comment (ngreenwald) [inline ark/utils/plot_utils.py:374]: This is very nice @ackagel, I had to make a hacky version of this for the mesmer paper. This will be useful

**Review by ngreenwald — CHANGES_REQUESTED** [review]
Looks good, just some small tweaks

File: templates/example_fiber_segmentation.ipynb [hunk 5]
```diff
@@ -0,0 +1,190 @@
+{
```
Comment (ngreenwald) [inline templates/example_fiber_segmentation.ipynb:1]: <p>This should be updated for the new language we have about mounting external dirs etc. </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/angelolab/ark-analysis/pull/569/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='2'/>

Comment (ngreenwald) [inline templates/example_fiber_segmentation.ipynb:1]: <p>We should run this with the example dataset so the use can see what the output looks like</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/angelolab/ark-analysis/pull/569/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='8'/>

Comment (ngreenwald) [inline templates/example_fiber_segmentation.ipynb:1]: <p>Returning the labeled images currently doesn't serve a purpose, and will take up a lot of memory. Let's just remove that functionality</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/angelolab/ark-analysis/pull/569/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='10'/>

File: ark/segmentation/fiber_segmentation.py (hunk 3 above)
Comment (camisowers) [reply ark/segmentation/fiber_segmentation.py:283]: Nope, the fiber label images should always be outputted. 
Per Adam's notebook documentation: "The function saves the fiber object table as a csv, and the labeled images as tiff files. The function also returns these as variables, for continued processing/exploration by the user."

Not sure what exactly the plan was, but it you think that's note something users will need I can remove the return.

File: ark/segmentation/fiber_segmentation.py [hunk 6]
```diff
@@ -0,0 +1,298 @@
...
+        data_dir (str | PathLike):
+            Folder containing dataset
+        fiber_channel (str):
+            Channel for fiber segmentation, e.g collagen.
+        out_dir (str | PathLike):
+            Directory to save fiber object labels and table.
+        img_sub_folder (str | NoneType):
+            Image subfolder name in `data_dir`. If there is not subfolder, set this to None.
+        **kwargs:
+            Keyword arguments for `segment_fibers`
```
[the remark this answers is not in the document]
Comment (camisowers) [reply ark/segmentation/fiber_segmentation.py:147]: The kwargs passed are just used to change any of the default args in segment_fibers (blur, contrast_scaling_divisor, fiber_widths, ridge_cutoff, sobel_blur, min_fiber_size, object_properties, save_csv, debug).

**Review by ngreenwald — APPROVED** [review]
Looks good!

**Pull request merged**